File: _may_messenger_mobile_app/lib/data/services/event_queue_service.dart
import 'dart:async';
import 'dart:collection';
import 'package:flutter/foundation.dart';
import 'package:logger/logger.dart';

/// Base class for all events processed by the queue
abstract class AppEvent {
  final DateTime timestamp;
  final String eventId;

  AppEvent({DateTime? timestamp, String? eventId})
      : timestamp = timestamp ?? DateTime.now(),
        eventId = eventId ?? DateTime.now().millisecondsSinceEpoch.toString();

  String get type;
}

/// Message received via SignalR
class MessageReceivedEvent extends AppEvent {
  final Map<String, dynamic> messageData;
  final String chatId;

  MessageReceivedEvent({
    required this.messageData,
    required this.chatId,
    super.timestamp,
    super.eventId,
  });

  @override
  String get type => 'MessageReceived';
}

/// Message sent successfully via REST API
class MessageSentEvent extends AppEvent {
  final String messageId;
  final String chatId;
  final String? clientMessageId;

  MessageSentEvent({
    required this.messageId,
    required this.chatId,
    this.clientMessageId,
    super.timestamp,
    super.eventId,
  });

  @override
  String get type => 'MessageSent';
}

/// Message status update (delivered, read, played)
class StatusUpdateEvent extends AppEvent {
  final String messageId;
  final int newStatus;
  final String source; // 'SignalR', 'REST', 'Local'

  StatusUpdateEvent({
    required this.messageId,
    required this.newStatus,
    required this.source,
    super.timestamp,
    super.eventId,
  });

  @override
  String get type => 'StatusUpdate';
}

/// Sync completed event
class SyncCompleteEvent extends AppEvent {
  final String chatId;
  final int messageCount;

  SyncCompleteEvent({
    required this.chatId,
    required this.messageCount,
    super.timestamp,
    super.eventId,
  });

  @override
  String get type => 'SyncComplete';
}

/// User status changed (online/offline)
class UserStatusChangedEvent extends AppEvent {
  final String userId;
  final bool isOnline;
  final DateTime lastSeenAt;

  UserStatusChangedEvent({
    required this.userId,
    required this.isOnline,
    required this.lastSeenAt,
    super.timestamp,
    super.eventId,
  });

  @override
  String get type => 'UserStatusChanged';
}

/// Typing indicator event
class TypingIndicatorEvent extends AppEvent {
  final String chatId;
  final String userId;
  final bool isTyping;

  TypingIndicatorEvent({
    required this.chatId,
    required this.userId,
    required this.isTyping,
    super.timestamp,
    super.eventId,
  });

  @override
  String get type => 'TypingIndicator';
}

/// Centralized event queue service for processing all app events sequentially
/// Ensures deduplication and ordered processing
class EventQueueService {
  final Logger _logger = Logger();
  final Queue<AppEvent> _queue = Queue();
  final Set<String> _processedEventIds = {};
  bool _processing = false;

  // Event handlers
  final Map<String, List<Function(AppEvent)>> _handlers = {};

  // Statistics
  int _totalEventsProcessed = 0;
  int _duplicatesSkipped = 0;
  DateTime? _lastProcessedAt;

  // Configuration
  static const int maxProcessedEventIdsSize = 1000;
  static const Duration eventTimeout = Duration(seconds: 30);

  EventQueueService() {
    _logger.d('[EventQueueService] Initialized');
  }

  /// Enqueue an event for processing
  void enqueue(AppEvent event) {
    // Check for duplicate
    if (_processedEventIds.contains(event.eventId)) {
      _duplicatesSkipped++;
      _logger.w('[EventQueueService] Duplicate event skipped: ${event.type} (${event.eventId})');
      return;
    }

    _queue.add(event);
    _logger.d('[EventQueueService] Event enqueued: ${event.type} (queue size: ${_queue.length})');

    // Start processing if not already running
    _processQueue();
  }

  /// Register a handler for a specific event type
  void registerHandler(String eventType, Function(AppEvent) handler) {
    if (!_handlers.containsKey(eventType)) {
      _handlers[eventType] = [];
    }
    _handlers[eventType]!.add(handler);
    _logger.d('[EventQueueService] Handler registered for event type: $eventType');
  }

  /// Unregister all handlers for a specific event type
  void unregisterHandlers(String eventType) {
    _handlers.remove(eventType);
    _logger.d('[EventQueueService] Handlers unregistered for event type: $eventType');
  }

  /// Process the queue sequentially
  Future<void> _processQueue() async {
    if (_processing) {
      return; // Already processing
    }

    _processing = true;

    try {
      while (_queue.isNotEmpty) {
        final event = _queue.removeFirst();

        // Check if event is too old (timeout)
        if (DateTime.now().difference(event.timestamp) > eventTimeout) {
          _logger.w('[EventQueueService] Event timeout: ${event.type} (${event.eventId})');
          continue;
        }

        // Process event
        await _processEvent(event);

        // Mark as processed
        _processedEventIds.add(event.eventId);
        _totalEventsProcessed++;
        _lastProcessedAt = DateTime.now();

        // Cleanup old processed event IDs to prevent memory leak
        if (_processedEventIds.length > maxProcessedEventIdsSize) {
          final toRemove = _processedEventIds.length - maxProcessedEventIdsSize;
          _processedEventIds.removeAll(_processedEventIds.take(toRemove));
        }
      }
    } catch (e, stackTrace) {
      _logger.e('[EventQueueService] Error processing queue', error: e, stackTrace: stackTrace);
    } finally {
      _processing = false;
    }
  }

  /// Process a single event
  Future<void> _processEvent(AppEvent event) async {
    try {
      _logger.d('[EventQueueService] Processing event: ${event.type} (${event.eventId})');

      final handlers = _handlers[event.type];
      if (handlers == null || handlers.isEmpty) {
        _logger.w('[EventQueueService] No handlers registered for event type: ${event.type}');
        return;
      }

      // Call all registered handlers
      for (final handler in handlers) {
        try {
          await handler(event);
        } catch (e, stackTrace) {
          _logger.e('[EventQueueService] Error in handler for ${event.type}', error: e, stackTrace: stackTrace);
        }
      }

      _logger.d('[EventQueueService] Event processed successfully: ${event.type}');
    } catch (e, stackTrace) {
      _logger.e('[EventQueueService] Error processing event: ${event.type}', error: e, stackTrace: stackTrace);
    }
  }

  /// Get queue statistics
  Map<String, dynamic> getStats() {
    return {
      'queueSize': _queue.length,
      'totalEventsProcessed': _totalEventsProcessed,
      'duplicatesSkipped': _duplicatesSkipped,
      'lastProcessedAt': _lastProcessedAt?.toIso8601String(),
      'isProcessing': _processing,
      'processedEventIdsSize': _processedEventIds.length,
    };
  }

  /// Clear the queue (use with caution)
  void clear() {
    _queue.clear();
    _logger.w('[EventQueueService] Queue cleared');
  }

  /// Dispose resources
  void dispose() {
    _queue.clear();
    _processedEventIds.clear();
    _handlers.clear();
    _logger.d('[EventQueueService] Disposed');
  }
}

File: _may_messenger_web_client/src/services/eventQueueService.ts
/**
 * Event Queue Service - обрабатывает все события приложения последовательно
 * Обеспечивает дедупликацию и упорядоченную обработку событий
 */

// Базовый класс для всех событий
export abstract class AppEvent {
  timestamp: Date;
  eventId: string;

  constructor(timestamp?: Date, eventId?: string) {
    this.timestamp = timestamp || new Date();
    this.eventId = eventId || Date.now().toString() + Math.random().toString(36);
  }

  abstract get type(): string;
}

// Событие получения сообщения через SignalR
export class MessageReceivedEvent extends AppEvent {
  constructor(
    public messageData: any,
    public chatId: string,
    timestamp?: Date,
    eventId?: string
  ) {
    super(timestamp, eventId);
  }

  get type(): string {
    return 'MessageReceived';
  }
}

// Событие успешной отправки сообщения через REST API
export class MessageSentEvent extends AppEvent {
  constructor(
    public messageId: string,
    public chatId: string,
    public clientMessageId?: string,
    timestamp?: Date,
    eventId?: string
  ) {
    super(timestamp, eventId);
  }

  get type(): string {
    return 'MessageSent';
  }
}

// Событие обновления статуса сообщения (delivered, read, played)
export class StatusUpdateEvent extends AppEvent {
  constructor(
    public messageId: string,
    public newStatus: number,
    public source: 'SignalR' | 'REST' | 'Local',
    timestamp?: Date,
    eventId?: string
  ) {
    super(timestamp, eventId);
  }

  get type(): string {
    return 'StatusUpdate';
  }
}

// Событие завершения синхронизации
export class SyncCompleteEvent extends AppEvent {
  constructor(
    public chatId: string,
    public messageCount: number,
    timestamp?: Date,
    eventId?: string
  ) {
    super(timestamp, eventId);
  }

  get type(): string {
    return 'SyncComplete';
  }
}

// Событие изменения статуса пользователя (online/offline)
export class UserStatusChangedEvent extends AppEvent {
  constructor(
    public userId: string,
    public isOnline: boolean,
    public lastSeenAt: Date,
    timestamp?: Date,
    eventId?: string
  ) {
    super(timestamp, eventId);
  }

  get type(): string {
    return 'UserStatusChanged';
  }
}

// Событие индикатора набора текста
export class TypingIndicatorEvent extends AppEvent {
  constructor(
    public chatId: string,
    public userId: string,
    public isTyping: boolean,
    timestamp?: Date,
    eventId?: string
  ) {
    super(timestamp, eventId);
  }

  get type(): string {
    return 'TypingIndicator';
  }
}

type EventHandler = (event: AppEvent) => void | Promise<void>;

/**
 * Централизованный сервис очереди событий для последовательной обработки всех событий приложения
 * Обеспечивает дедупликацию и упорядоченную обработку
 */
class EventQueueService {
  private queue: AppEvent[] = [];
  private processedEventIds = new Set<string>();
  private processing = false;

  // Обработчики событий
  private handlers = new Map<string, EventHandler[]>();

  // Статистика
  private totalEventsProcessed = 0;
  private duplicatesSkipped = 0;
  private lastProcessedAt: Date | null = null;

  // Конфигурация
  private static readonly MAX_PROCESSED_EVENT_IDS_SIZE = 1000;
  private static readonly EVENT_TIMEOUT = 30000; // 30 секунд

  constructor() {
    console.log('[EventQueueService] Initialized');
  }

  /**
   * Добавить событие в очередь для обработки
   */
  enqueue(event: AppEvent): void {
    // Проверка на дубликат
    if (this.processedEventIds.has(event.eventId)) {
      this.duplicatesSkipped++;
      console.log(`[EventQueueService] Duplicate event skipped: ${event.type} (${event.eventId})`);
      return;
    }

    this.queue.push(event);
    console.log(`[EventQueueService] Event enqueued: ${event.type} (queue size: ${this.queue.length})`);

    // Запустить обработку, если еще не запущена
    this.processQueue();
  }

  /**
   * Зарегистрировать обработчик для конкретного типа события
   */
  registerHandler(eventType: string, handler: EventHandler): void {
    if (!this.handlers.has(eventType)) {
      this.handlers.set(eventType, []);
    }
    this.handlers.get(eventType)!.push(handler);
    console.log(`[EventQueueService] Handler registered for event type: ${eventType}`);
  }

  /**
   * Удалить все обработчики для конкретного типа события
   */
  unregisterHandlers(eventType: string): void {
    this.handlers.delete(eventType);
    console.log(`[EventQueueService] Handlers unregistered for event type: ${eventType}`);
  }

  /**
   * Обработать очередь последовательно
   */
  private async processQueue(): Promise<void> {
    if (this.processing) {
      return; // Уже обрабатывается
    }

    this.processing = true;

    try {
      while (this.queue.length > 0) {
        const event = this.queue.shift()!;

        // Проверка на таймаут события (слишком старое)
        if (Date.now() - event.timestamp.getTime() > EventQueueService.EVENT_TIMEOUT) {
          console.log(`[EventQueueService] Event timeout: ${event.type} (${event.eventId})`);
          continue;
        }

        // Обработать событие
        await this.processEvent(event);

        // Пометить как обработанное
        this.processedEventIds.add(event.eventId);
        this.totalEventsProcessed++;
        this.lastProcessedAt = new Date();

        // Очистка старых обработанных ID событий для предотвращения утечки памяти
        if (this.processedEventIds.size > EventQueueService.MAX_PROCESSED_EVENT_IDS_SIZE) {
          const toRemove = this.processedEventIds.size - EventQueueService.MAX_PROCESSED_EVENT_IDS_SIZE;
          const iterator = this.processedEventIds.values();
          for (let i = 0; i < toRemove; i++) {
            this.processedEventIds.delete(iterator.next().value);
          }
        }
      }
    } catch (error) {
      console.error('[EventQueueService] Error processing queue:', error);
    } finally {
      this.processing = false;
    }
  }

  /**
   * Обработать одно событие
   */
  private async processEvent(event: AppEvent): Promise<void> {
    try {
      console.log(`[EventQueueService] Processing event: ${event.type} (${event.eventId})`);

      const handlers = this.handlers.get(event.type);
      if (!handlers || handlers.length === 0) {
        console.log(`[EventQueueService] No handlers registered for event type: ${event.type}`);
        return;
      }

      // Вызвать все зарегистрированные обработчики
      for (const handler of handlers) {
        try {
          await handler(event);
        } catch (error) {
          console.error(`[EventQueueService] Error in handler for ${event.type}:`, error);
        }
      }

      console.log(`[EventQueueService] Event processed successfully: ${event.type}`);
    } catch (error) {
      console.error(`[EventQueueService] Error processing event: ${event.type}:`, error);
    }
  }

  /**
   * Получить статистику очереди
   */
  getStats(): {
    queueSize: number;
    totalEventsProcessed: number;
    duplicatesSkipped: number;
    lastProcessedAt: string | null;
    isProcessing: boolean;
    processedEventIdsSize: number;
  } {
    return {
      queueSize: this.queue.length,
      totalEventsProcessed: this.totalEventsProcessed,
      duplicatesSkipped: this.duplicatesSkipped,
      lastProcessedAt: this.lastProcessedAt?.toISOString() || null,
      isProcessing: this.processing,
      processedEventIdsSize: this.processedEventIds.size,
    };
  }

  /**
   * Очистить очередь (использовать с осторожностью)
   */
  clear(): void {
    this.queue = [];
    console.log('[EventQueueService] Queue cleared');
  }

  /**
   * Освободить ресурсы
   */
  dispose(): void {
    this.queue = [];
    this.processedEventIds.clear();
    this.handlers.clear();
    console.log('[EventQueueService] Disposed');
  }
}

// Экспортируем синглтон
export const eventQueueService = new EventQueueService();

